File: java/src/main/java/benchmarks/Mandelbrot.java
package benchmarks;

import java.io.ByteArrayOutputStream;

public class Mandelbrot extends Benchmark {
    private static final int ITER = 50;
    private static final double LIMIT = 2.0;

    private int w, h;
    private ByteArrayOutputStream result;

    public Mandelbrot() {
        w = (int) configVal("w");
        h = (int) configVal("h");
        result = new ByteArrayOutputStream();
    }

    @Override
    public String name() {
        return "CLBG::Mandelbrot";
    }

    @Override
    public void run(int iterationId) {
        try {
            result.write(("P4\n" + w + " " + h + "\n").getBytes());

            int bitNum = 0;
            byte byteAcc = 0;

            for (int y = 0; y < h; y++) {
                for (int x = 0; x < w; x++) {
                    double zr = 0.0, zi = 0.0;
                    double tr = 0.0, ti = 0.0;
                    double cr = (2.0 * x / w - 1.5);
                    double ci = (2.0 * y / h - 1.0);

                    int i = 0;
                    while (i < ITER && (tr + ti <= LIMIT * LIMIT)) {
                        zi = 2.0 * zr * zi + ci;
                        zr = tr - ti + cr;
                        tr = zr * zr;
                        ti = zi * zi;
                        i++;
                    }

                    byteAcc <<= 1;
                    if (tr + ti <= LIMIT * LIMIT) {
                        byteAcc |= 0x01;
                    }
                    bitNum++;

                    if (bitNum == 8) {
                        result.write(byteAcc);
                        byteAcc = 0;
                        bitNum = 0;
                    } else if (x == w - 1) {
                        byteAcc <<= (8 - (w % 8));
                        result.write(byteAcc);
                        byteAcc = 0;
                        bitNum = 0;
                    }
                }
            }
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }

    @Override
    public long checksum() {
        return Helper.checksum(result.toByteArray());
    }
}

File: scala/src/main/scala/benchmarks/Mandelbrot.scala
package benchmarks

import java.io.ByteArrayOutputStream

class Mandelbrot extends Benchmark:
  private val ITER = 50
  private val LIMIT = 2.0

  private var w: Int = 0
  private var h: Int = 0
  private val result = new ByteArrayOutputStream()

  override def name(): String = "CLBG::Mandelbrot"

  override def prepare(): Unit =
    w = configVal("w").toInt
    h = configVal("h").toInt

  override def run(iterationId: Int): Unit =
    result.write(s"P4\n$w $h\n".getBytes)

    var bitNum = 0
    var byteAcc: Byte = 0

    var y = 0
    while y < h do
      var x = 0
      while x < w do
        var zr = 0.0
        var zi = 0.0
        var tr = 0.0
        var ti = 0.0
        val cr = (2.0 * x / w - 1.5)
        val ci = (2.0 * y / h - 1.0)

        var i = 0
        while i < ITER && (tr + ti <= LIMIT * LIMIT) do
          zi = 2.0 * zr * zi + ci
          zr = tr - ti + cr
          tr = zr * zr
          ti = zi * zi
          i += 1

        byteAcc = (byteAcc << 1).toByte
        if tr + ti <= LIMIT * LIMIT then byteAcc = (byteAcc | 0x01).toByte
        bitNum += 1

        if bitNum == 8 then
          result.write(byteAcc)
          byteAcc = 0
          bitNum = 0
        else if x == w - 1 then
          byteAcc = (byteAcc << (8 - (w % 8))).toByte
          result.write(byteAcc)
          byteAcc = 0
          bitNum = 0

        x += 1
      y += 1

  override def checksum(): Long =
    Helper.checksum(result.toByteArray())
